### ISSUE TITLE
Higher order derivatives of norm.logcdf seem to have numerical problems at low input values

### ISSUE THREAD

**Issue by martiningram** [google/jax#694]
Hi there,

I think I might have found a problem while investigating https://github.com/google/jax/issues/475 (though still checking!).

It looks like second derivatives of the logcdf and higher have problems for small input values (<-10) -- see reproducible code below:

```python
import os
os.environ['XLA_FLAGS'] = '--xla_cpu_enable_fast_math=false'
import jax.numpy as np
from jax.scipy.stats import norm
from jax import grad, vmap

x_vals = np.linspace(-30, 20, 100)
first_grad = vmap(grad(norm.logcdf))(x_vals)
second_grad = vmap(grad(grad(norm.logcdf)))(x_vals)
third_grad = vmap(grad(grad(grad(norm.logcdf))))(x_vals)

count_nans = lambda x: int(np.sum(np.isnan(x)))

print(f'Found {count_nans(first_grad)} NaNs in first derivative.')
print(f'Found {count_nans(second_grad)} NaNs in second derivative.')
print(f'Found {count_nans(third_grad)} NaNs in third derivative.')
```

Output is:

```
Found 0 NaNs in first derivative.
Found 40 NaNs in second derivative.
Found 42 NaNs in third derivative.
```

I need derivatives up to third order and have been able to get around the issue by defining the ratio:

```python
def pdf_cdf_ratio(x):
    return np.exp(norm.logpdf(x) - norm.logcdf(x))
```

Which is the gradient of the logcdf (subtracting and then exponentiating is much more stable than computing the ratio directly). You can express the second and third derivatives in terms of the ratio. Here's the full code:

```python
def pdf_cdf_ratio(x):
    return np.exp(norm.logpdf(x) - norm.logcdf(x))

def pdf_cdf_ratio_grad(x):

    g = pdf_cdf_ratio(x)
    return -g * (x + g)

def pdf_cdf_ratio_grad_grad(x):

    g = pdf_cdf_ratio(x)
    gp = pdf_cdf_ratio_grad(x)

    return -gp * (x + g) - g * (1 + gp)
```

With these alternatives, I get:

```python
first_grad_alt = pdf_cdf_ratio(x_vals)
second_grad_alt = pdf_cdf_ratio_grad(x_vals)
third_grad_alt = pdf_cdf_ratio_grad_grad(x_vals)

count_nans = lambda x: int(np.sum(np.isnan(x)))

print(f'Found {count_nans(first_grad_alt)} NaNs in first derivative.')
print(f'Found {count_nans(second_grad_alt)} NaNs in second derivative.')
print(f'Found {count_nans(third_grad_alt)} NaNs in third derivative.')
```

```
Found 0 NaNs in first derivative.
Found 0 NaNs in second derivative.
Found 0 NaNs in third derivative.
```

Mind you, plotting the functions, although it has no `nan`s, it looks like the third gradient gets pretty wonky below values of -10, so maybe there's a better alternative.

Anyway, I'm not sure what to do to fix this with autodiff, and whether that's even possible, but thought I'd raise this as a possible problem!

**Comment by martiningram** [comment]
It may also be of interest (apologies if not!) that autograd doesn't have this problem:

```python
import autograd.numpy as np
from autograd.scipy.stats import norm
from autograd import elementwise_grad

inputs = np.linspace(-50, 20, 100)

fun = norm.logcdf
grad_fun = elementwise_grad(fun)
grad_grad_fun = elementwise_grad(grad_fun)
grad_grad_grad_fun = elementwise_grad(grad_grad_fun)

f = fun(inputs)
fp = grad_fun(inputs)
fpp = grad_grad_fun(inputs)
fppp = grad_grad_grad_fun(inputs)

[np.any(np.isnan(x)) for x in [f, fp, fpp, fppp]]
```

Gives:

```
[False, False, False, False]
```

And the plots actually look nicer than what my ratio trick gives, so it seems to have a nice solution there. Looks like there, the vjp is defined in terms of the ratio:

https://github.com/HIPS/autograd/blob/master/autograd/scipy/stats/norm.py

Sorry if this was all obvious already, but I thought I'd add this information in case it's helpful!

**Comment by hawkinsp** [comment]
Our guess is that differentiating the (rather complicated!) implementation of `log_ndtr` isn't working that well, and that problem can be fixed by adding a derivative that treats `log_ndtr` as a black box, much like autograd does.
